recipes-kernel/linux/4.19/patches/blktap2.patch

```diff
@@ -0,0 +1,3066 @@
++blktap_sysfs_get_name(struct device *dev, struct device_attribute *attr, char *buf)
++{
++	struct blktap *tap;
++	ssize_t size;
++
++	tap = dev_get_drvdata(dev);
++	if (!tap)
++		return 0;
++
++	if (tap->name[0])
++		size = sprintf(buf, "%s\n", tap->name);
++	else
++		size = sprintf(buf, "%d\n", tap->minor);
++
++	return size;
++}
++static DEVICE_ATTR(name, S_IRUGO|S_IWUSR,
++		   blktap_sysfs_get_name, blktap_sysfs_set_name);
++
++static void
++blktap_sysfs_remove_work(struct work_struct *work)
++{
++	struct blktap *tap
++		= container_of(work, struct blktap, remove_work);
++	blktap_control_destroy_tap(tap);
++}
++
++static ssize_t
++blktap_sysfs_remove_device(struct device *dev,
++			   struct device_attribute *attr,
++			   const char *buf, size_t size)
++{
++	struct blktap *tap;
++	int err;
++
++	tap = dev_get_drvdata(dev);
++	if (!tap)
++		return size;
++
++	if (test_and_set_bit(BLKTAP_SHUTDOWN_REQUESTED, &tap->dev_inuse))
++		goto wait;
++
++	if (tap->ring.vma) {
++		struct blktap_sring *sring = tap->ring.ring.sring;
++		*BLKTAP_RING_MESSAGE(sring) = BLKTAP_RING_MESSAGE_CLOSE;
++		blktap_ring_kick_user(tap);
++	} else {
++		INIT_WORK(&tap->remove_work, blktap_sysfs_remove_work);
++		schedule_work(&tap->remove_work);
++	}
++wait:
++	err = wait_event_interruptible(tap->remove_wait,
++				       !dev_get_drvdata(dev));
++	if (err)
++		return err;
++
++	return size;
++}
++static DEVICE_ATTR(remove, S_IWUSR, NULL, blktap_sysfs_remove_device);
++
++static ssize_t
++blktap_sysfs_pause_device(struct device *dev, struct device_attribute *attr,
++                          const char *buf, size_t size)
++{
++        int err;
++        struct blktap *tap = dev_get_drvdata(dev);
++
++        BTDBG("pausing %u:%u: dev_inuse: %lu\n",
++              MAJOR(tap->ring.devno), MINOR(tap->ring.devno), tap->dev_inuse);
++
++        if (!tap->ring.dev ||
++            test_bit(BLKTAP_SHUTDOWN_REQUESTED, &tap->dev_inuse)) {
++                err = -ENODEV;
++                goto out;
++        }
++
++        if (test_bit(BLKTAP_PAUSE_REQUESTED, &tap->dev_inuse)) {
++                err = -EBUSY;
++                goto out;
++        }
++
++        if (test_bit(BLKTAP_PAUSED, &tap->dev_inuse)) {
++                err = 0;
++                goto out;
++        }
++
++        err = blktap_device_pause(tap);
++
++out:
++
++        return (err ? err : size);
++}
++static DEVICE_ATTR(pause, S_IWUSR, NULL, blktap_sysfs_pause_device);
++
++static ssize_t
++blktap_sysfs_resume_device(struct device *dev, struct device_attribute *attr,
++                           const char *buf, size_t size)
++{
++	int err;
++	struct blktap *tap = dev_get_drvdata(dev);
++
++	if (!tap->ring.dev ||
++            test_bit(BLKTAP_SHUTDOWN_REQUESTED, &tap->dev_inuse)) {
++		err = -ENODEV;
++		goto out;
++        }
++
++	if (!test_bit(BLKTAP_PAUSED, &tap->dev_inuse)) {
++		err = -EINVAL;
++		goto out;
++        }
++
++	err = blktap_device_resume(tap);
++
++out:
++
++	BTDBG("returning %d\n", (
```

Comment:
Re-exporting elevator_init is easiest for now, but is probably wrong long term.  The linux block subsystem will be multiqueue-only, and blk_init_queue will be going away in addition to elevator_init.  I don't know what it would take to convert multiqueue, but I imagine that will need to be done.

Reply:
Agreed it will have to be done. I took a few hours last week and determined its probably going to take a week of dev/test work on the blktap driver to get it working with multi queue. This would be a good follow-on task once we rebase and accept the upstream blktap implementation.